File: src/main/java/com/kingmang/ixion/ast/BinaryExpression.java
package com.kingmang.ixion.ast;

import com.kingmang.ixion.ExprVisitor;
import com.kingmang.ixion.lexer.Position;
import com.kingmang.ixion.lexer.Token;

public final class BinaryExpression extends Expression {
    public final Expression left;
    public final Token operator;
    public final Expression right;

    public BinaryExpression(Position pos, Expression left, Token operator, Expression right) {
        super(pos);
        this.left = left;
        this.operator = operator;
        this.right = right;
    }

    public <R> R accept(ExprVisitor<R> visitor) {
        return visitor.visitBinaryExpr(this);
    }
}

File: src/main/java/com/kingmang/ixion/ast/BinaryExpression.kt
package com.kingmang.ixion.ast

import com.kingmang.ixion.ExprVisitor
import com.kingmang.ixion.lexer.Position
import com.kingmang.ixion.lexer.Token

class BinaryExpression(pos: Position?, @JvmField val left: Expression?, @JvmField val operator: Token?, @JvmField val right: Expression?) :
    Expression(pos) {
    override fun <R> accept(visitor: ExprVisitor<R>): R {
        return visitor.visitBinaryExpr(this)
    }
}